### PR TITLE
Add autokey

### PR DESCRIPTION
**Pull request by michsa**
WIP because it's dependent on smartprocure/futil-js#270, but I expect there will be plenty of feedback on this one, so I wanted to get the PR out anyway now that the rest is finished.

I definitely left the futil methods in there so CI will still cough up my hard-earned green checkmark -- but also, hopefully seeing it all together will better communicate what I was trying to go for with the `uniqueStringHash` stuff. It was fun and all, but I am far from convinced that it's actually any good. ¯\\\_(ツ)\_/¯

### CODE REVIEW

**Comment by decrapifier** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>Could not find any browser results.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>We were able to automatically fix some formatting issues in this PR for you!</td>
    </tr>
  </tbody>
</table>


### Some things that were possibly fixed:

- Code that could be fixed via the --fix flag
- Formatting that could be fixed by prettier

Take a look at this commit to see what happened in detail: b5b5ae0f1fd5eb4169c11e254086c387db45801d


And look at this wiki page to see the reasoning behind the ESLint rules: https://github.com/smartprocure/eslint-config-smartprocure/wiki/Rules-and-Why-We-Chose-Them
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 28a3a301a69ffd43e1ab1f354f415f291c7b373d
</p>

**Comment by michsa** [comment]
Now includes futil@1.58.0 dependency (which does not actually exist at the time of writing)

File: src/node.js [hunk 1]
```diff
@@ -22,16 +23,30 @@ export let internalStateKeys = {
...
   runTypeFunction(types, 'init', node, extend)
-  extend(node, {
-    ..._.omit(_.keys(node), defaults),
-    ..._.omit(_.keys(node), getTypeProp(types, 'defaults', node)),
-    path,
-  })
+  if (node.key) dedupe(node.key) // add node.key to the dedupe cache
+  let key =
+    node.key ||
+    dedupe(runTypeFunctionOrDefault(autoKey, types, 'autoKey', node, extend))
```
Comment (daedalus28) [inline src/node.js]: we probably could have just reused `getTypeProp` instead of creating `runTypeFunctionOrDefault`

File: src/node.js [hunk 2]
```diff
@@ -22,16 +23,30 @@ export let internalStateKeys = {
...
+  extend(
+    node,
+    _.defaults(
+      { ...defaults, ...getTypeProp(types, 'defaults', node) },
+      { ...node, key, path: [...parentPath, key] }
```
Comment (daedalus28) [inline src/node.js]: This change which gets rid of the `omit` and instead spreads node will cause it to trigger observable updates for every single property on every single node which is not ideal

File: src/actions/index.js [hunk 3]
```diff
@@ -28,14 +28,27 @@ export default config => {
...
-      initNode(node, path, extend, types)
-      flat[encode(path)] = node
-    })(node)
-
     let target = getNode(parentPath)
+    // initialize uniqueString cache for the parent of the node to be added here,
+    // since it's not visited during the tree walk
+    let parentDedupeChildren = F.uniqueString(_.map('key', target.children))
+    node = initObject(node)
+    Tree.walk(
```
Comment (daedalus28) [inline src/actions/index.js]: The bulk of this walk is repeated here and in the index. We should probably factor something out for this

File: src/node.js (hunk 1 above)
Comment (michsa) [reply src/node.js]: Doing this with `getTypeProp` basically comes out to `(getTypeProp(types, 'autokey', node) || autoKey)(node)` -- which is just the body of `runTypeFunction` except with `autoKey` as the default function instead of `_.stubTrue`.

It was such a trivial change to generalize it to `runTypeFunctionOrDefault` that even with only one use, I still think it ends up cleaner than the `getTypeProp` alternative

File: src/node.js [hunk 4]
```diff
@@ -22,16 +24,37 @@ export let internalStateKeys = {
...
   })
+})
+
+// fn: (dedupe: string -> string, parentPath: array, node: object) -> void
+export let dedupeWalk = (fn, tree, { target = {}, dedupe } = {}) => {
+  // allows us to maintain separate deduplication caches for each node's children by
+  // storing a `uniqueString` instance in `dedupes` with that node's path as its key.
+  // this ensures that autogenerated node keys will always be unique from their siblings,
+  // but won't be unnecessarily modified if they are duplicates of keys in other branches.
+  let dedupes = target.path ? { [encode(target.path)]: dedupe } : {}
```
Comment (michsa) [inline src/node.js]: If the shoe fits...

**Comment by daedalus28** [comment]
Let's add some more tests to cover all the edge cases we're handling now
